src/Temporalio/Worker/ActivityWorker.cs

```diff
@@ -0,0 +1,703 @@
+            {
+                ["TaskQueue"] = worker.Options.TaskQueue!,
+            }))
+            {
+                while (true)
+                {
+                    var task = await worker.BridgeWorker.PollActivityTaskAsync().ConfigureAwait(false);
+                    logger.LogTrace("Received activity task: {Task}", task);
+                    switch (task?.VariantCase)
+                    {
+                        case Bridge.Api.ActivityTask.ActivityTask.VariantOneofCase.Start:
+                            StartActivity(task);
+                            break;
+                        case Bridge.Api.ActivityTask.ActivityTask.VariantOneofCase.Cancel:
+                            if (!runningActivities.TryGetValue(task.TaskToken, out var act))
+                            {
+                                logger.LogWarning(
+                                    "Cannot find activity to cancel for token {TaskToken}",
+                                    task.TaskToken);
+                            }
+                            else
+                            {
+                                act.Cancel(task.Cancel.Reason);
+                            }
+                            break;
+                        case null:
+                            // This means worker shut down
+                            return;
+                        default:
+                            throw new InvalidOperationException(
+                                $"Unexpected activity task case {task?.VariantCase}");
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gracefully shutdown this worker after <see cref="ExecuteAsync" /> is complete. This must
+        /// not be called until <see cref="ExecuteAsync" /> has returned. This should never throw.
+        /// </summary>
+        /// <returns>
+        /// Task that completes when all activities have completed. If a running activity does not
+        /// respond to cancellation, this may never return.
+        /// </returns>
+        public async Task GracefulShutdownAsync()
+        {
+            using (logger.BeginScope(new Dictionary<string, object>()
+            {
+                ["TaskQueue"] = worker.Options.TaskQueue!,
+            }))
+            {
+                workerShutdownTokenSource.Cancel();
+                logger.LogInformation(
+                    "Beginning activity worker shutdown, will wait {GracefulShutdownTimeout} before " +
+                    "cancelling {ActivityCount} activity instance(s)",
+                    worker.Options.GracefulShutdownTimeout,
+                    runningActivities.Count);
+                await Task.Delay(worker.Options.GracefulShutdownTimeout).ConfigureAwait(false);
+                // Issue cancel-all and wait for all running activities to complete
+                await Task.WhenAll(runningActivities.Values.Select(act =>
```

Comment:
You should wait for poll to return shutdown error to ensure that no activities are added to this dictionary before this line.

Looks like there's no guarantee here: https://github.com/temporalio/sdk-dotnet/pull/18/files/b1eb2174cda30c4110f0dde22a800fe4d7636535#diff-49d90da0465bf6446c00276cd2c7e7f5e4832c01ddd1c426a6af81b1b689ff53R260

Reply:
Yes, I mention at the top of this call that it should not be called until "ExecuteAsync" is done but I didn't ensure that from the caller. Fixing.

Reply:
Fixed